RedHatInsights/rhsm-subscriptions
# SWATCH-2522: Skip syncing subscriptionUMB messages for newly created contracts

Jira issue: SWATCH-2522

## Description

Subscription UMB messages are sent along side contract UMB messages when a new contract is created. The subscription UMB message has an inaccurate start_date that we should ignore in favor of the more precise contract start_date. This change prevents the initial UMB subscription message from creating a subscription with the wrong start_date.
## Testing
This requires testing in stage. We need to create new aws and azure contracts and test updating and inactivating the contracts.

The liquibase can be tested as follows:
**Setup fresh database off of main if needed.**
1. git checkout main
2. podman-compose down
3. podman-compose up -d
4. ./gradlew :bootRun

**Checkout branch, insert data, and start app to run liquibase**
5. git checkout kflahert/swatch-2522-duplicate-payg-subscriptions
6. Insert test data:
```
INSERT INTO public.offering VALUES ('RH02781HR', 'RHEL Server', 'RHEL', NULL, NULL, NULL, NULL, NULL, 'Premium', 'Production', 'Red Hat Enterprise Linux for Third Party Linux Migration with Extended Life Cycle Support (On-Demand, Hourly Production Support)', false, NULL, false);
INSERT INTO public.offering VALUES ('MW02393', 'OpenShift Container Platform', NULL, 0, 0, 0, 0, 'ocp', 'Premium', 'Production', 'Red Hat OpenShift Container Platform (Hourly)', NULL, NULL, NULL);

INSERT INTO public.subscription VALUES ('MW02393', 'org123', '12345', 1, '2023-08-01 04:00:00+00', '2024-08-01 03:59:59+00', 'testing', '12222', 'aws', 'testing');
INSERT INTO public.subscription VALUES ('MW02393', 'org123', '12345', 1, '2023-08-01 18:08:41.764447+00', '2023-09-21 15:06:29.374969+00', NULL, '12222', 'aws', 'testing');
INSERT INTO public.subscription VALUES ('RH02781HR', 'org456', '6789', 1, '2024-04-26 04:00:00+00', '2025-04-27 03:59:59.999+00', 'testing', '13333', 'aws', 'testing');
INSERT INTO public.subscription VALUES ('RH02781HR', 'org456', '6789', 1, '2024-04-26 12:10:41.076776+00', NULL, 'testing', '13333', 'aws', 'testing');
INSERT INTO public.subscription_product_ids VALUES ('rosa', '12345', '2023-08-01 04:00:00+00');
INSERT INTO public.subscription_measurements VALUES ('12345', '2023-08-01 04:00:00+00', 'Sockets', 'PHYSICAL', 4);

-- Control group - Should not be affected
INSERT INTO public.subscription VALUES ('MW02393', 'org456', '11111', 1, '2024-05-07 11:47:53.315975+00', '2025-05-07 03:59:59.999+00', 'testing', '14444', 'aws', 'testing');
INSERT INTO public.subscription VALUES ('MW02393', 'org123', '22222', 1, '2024-05-08 08:18:33.440393+00', '2025-05-08 03:59:59.999+00', NULL, '15555', NULL, NULL);
INSERT INTO public.subscription VALUES ('MW02393', 'org123', '22222', 1, '2024-05-08 05:00:00+00', '2025-02-13 04:59:59+00', NULL, '15555', NULL, NULL);
INSERT INTO public.subscription_product_ids VALUES ('rosa', '22222', '2024-05-08 05:00:00+00');
INSERT INTO public.subscription_measurements VALUES ('22222', '2024-05-08 05:00:00+00', 'Sockets', 'PHYSICAL', 266);
```
7. Start the app with `DEV_MODE=true ./gradlew :bootRun`
8. Verify that 5 subscriptions remain and one subscription_product_ids record and subscription_measurements record.


### Testing nightly sync:
1. Insert test data:
```
INSERT INTO public.org_config VALUES ('123', 'API', '2024-05-22 00:00:00+00', '2024-05-22 00:00:00+00');
INSERT INTO public.subscription VALUES ('MW01882', '123', '235252', 1, '2011-01-01 13:19:33+00', '2031-01-01 01:02:33+00', 'oldId', '4243626', 'aws', '1234567891234');
```
2. Start app with: `DEV_MODE=true SUBSCRIPTION_USE_STUB=true SUBSCRIPTION_SYNC_ENABLED=true PRODUCT_USE_STUB=true ./gradlew :bootRun`
3. Run curl: `curl -X 'PUT'   "localhost:8000/api/rhsm-subscriptions/v1/internal/rpc/subscriptions/sync?forceSync=false"   -H 'accept: application/json'   -H 'x-rh-swatch-psk: placeholder'`
4. Check subscription record has same start date, new billing_account_id of 1234567891234 and no duplicate created for that subscription_number(4243626)
